### PR TITLE
feat(npm): add v3 lock file support

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
Add v3 lock file support - https://docs.npmjs.com/cli/v9/configuring-npm/package-lock-json?v=true#lockfileversion

### CODE REVIEW

File: pkg/nodejs/npm/parse.go [hunk 1]
```diff
@@ -53,13 +61,106 @@ func (p *Parser) Parse(r dio.ReadSeekerAt) ([]types.Library, []types.Dependency,
...
+				continue
+			}
+
+			// Try to resolve the version with the higher level dependencies
+			depPath = filepath.Join(nodeModulesFolder, depName)
+			if dep, ok := packages[depPath]; ok {
+				depID := utils.PackageID(depName, dep.Version)
+				dependsOn = append(dependsOn, depID)
+				continue
+			}
```
Comment (knqyf263) [inline pkg/nodejs/npm/parse.go:145]: Isn't it possible to have more than two nests like `node_modules/body-parser/node_modules/debug/node_modules/ms`? Looks like it happens theoretically.

What about this code?

```suggestion
			depPath := path.Join(pkgPath, nodeModulesFolder)
			paths := strings.Split(depPath, "/")

			// Try to resolve the version with the nearest directory
			// In the case of "ms", try the following paths in order.
			//    - "node_modules/body-parser/node_modules/debug/node_moduls/ms"
			//    - "node_modules/body-parser/node_modules/ms"
			//    - "node_modules/ms"
			for i := len(paths) - 1; i >= 0; i-- {
				if paths[i] != nodeModulesFolder {
					continue
				}
				p := path.Join(paths[:i+1]...)
				p = path.Join(p, depName)
	
				if dep, ok := packages[p]; ok {
					depID := utils.PackageID(depName, dep.Version)
					dependsOn = append(dependsOn, depID)
					continue
				}
			}
```

https://go.dev/play/p/UPQN11T8qJ_m

File: pkg/nodejs/npm/parse.go [hunk 2]
```diff
@@ -137,12 +238,30 @@ func uniqueDeps(deps []types.Dependency) []types.Dependency {
...
+	if directVersion, ok := directDeps[libName]; ok {
+		// project can contain 2 different version of dependency
+		// we need to check that we choose version for direct dependency
+		// also the direct dependency version can use range
+		// e.g. "body-parser": "^1.18.3"
+		// we need to compare versions using npm comparator
+		match, err := matchVersion(directVersion, libVersionRange)
+		if err != nil {
+			log.Logger.Warnf("unable to compare version for %s@%s", libName, libVersionRange)
+		}
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/nodejs/npm/parse.go]: >  direct dependencies are always located just under node_modules.

You are right! i didn't think about this.
Changed this logic.

File: pkg/nodejs/npm/parse.go [hunk 3]
```diff
@@ -53,13 +61,106 @@ func (p *Parser) Parse(r dio.ReadSeekerAt) ([]types.Library, []types.Dependency,
...
 }
 
-func (p *Parser) parse(dependencies map[string]Dependency, dircetDeps map[string]string, versions map[string]string) ([]types.Library, []types.Dependency) {
+func (p *Parser) parseV2(packages map[string]Package) ([]types.Library, []types.Dependency) {
+	libs := make(map[string]types.Library, len(packages)-1)
+	var deps []types.Dependency
+
+	directDeps := map[string]string{}
+	for name, version := range packages[""].Dependencies {
+		pkgPath := filepath.Join(nodeModulesFolder, name)
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/nodejs/npm/parse.go]: Thanks for tip, fixed.

File: pkg/nodejs/npm/parse.go [hunk 4]
```diff
@@ -137,12 +238,30 @@ func uniqueDeps(deps []types.Dependency) []types.Dependency {
...
+		return !match
+	}
+	return true
+}
+
+func pkgNameFromPath(path string) string {
+	// lock file contains path to dependency in `node_modules` folder. e.g.:
+	// node_modules/string-width
+	// node_modules/string-width/node_modules/strip-ansi
+	return filepath.Base(path)
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/nodejs/npm/parse.go]: Fixed

File: pkg/nodejs/npm/parse.go (hunk 1 above)
Comment (DmitriyLewen) [reply pkg/nodejs/npm/parse.go]: Good idea! thanks, used this.

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**